File: app/src/main/java/com/example/recomendaciones_app/services/VentasService.java
package com.example.recomendaciones_app.services;

import android.util.Log;

import org.json.JSONObject;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * DEPRECATED: Implementación Java de notificador de ventas.
 * Se mantiene solo por compatibilidad pero la implementación recomendada
 * es la clase Kotlin `VentasService` en el mismo paquete.
 */
@Deprecated
class VentasServiceJava {

    // Cambia por tu webhook real (sin espacios)
    public static String WEBHOOK_URL = "https://primary-production-2873d.up.railway.app/webhook/nueva-venta";

    public interface VentaCallback {
        void onResult(boolean success, String message);
    }

    public static class VentaData {
        public int producto_id;
        public String producto_nombre;
        public int cantidad;
        public double precio_unitario;
        public double total;
        public String usuario_id; // nullable
        public String fecha; // yyyy-MM-dd HH:mm:ss
        public String metodo_pago;

        public VentaData(int producto_id, String producto_nombre, int cantidad, double precio_unitario,
                         double total, String usuario_id, String fecha, String metodo_pago) {
            this.producto_id = producto_id;
            this.producto_nombre = producto_nombre;
            this.cantidad = cantidad;
            this.precio_unitario = precio_unitario;
            this.total = total;
            this.usuario_id = usuario_id;
            this.fecha = fecha;
            this.metodo_pago = metodo_pago;
        }
    }

    public VentasServiceJava() {
    }

    public static String getCurrentFormattedDate() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
        return sdf.format(new Date());
    }

    /**
     * Envía la venta al webhook en background y reporta resultado por callback.
     */
    public void notificarVenta(final VentaData venta, final VentaCallback callback) {
        new Thread(() -> {
            HttpURLConnection conn = null;
            try {
                JSONObject payload = new JSONObject();
                payload.put("producto_id", venta.producto_id);
                payload.put("producto_nombre", venta.producto_nombre);
                payload.put("cantidad", venta.cantidad);
                payload.put("precio_unitario", venta.precio_unitario);
                payload.put("total", venta.total);
                payload.put("usuario_id", venta.usuario_id == null ? JSONObject.NULL : venta.usuario_id);
                payload.put("fecha", venta.fecha);
                payload.put("metodo_pago", venta.metodo_pago);

                URL url = new URL(WEBHOOK_URL);
                conn = (HttpURLConnection) url.openConnection();
                conn.setConnectTimeout(15000);
                conn.setReadTimeout(15000);
                conn.setRequestMethod("POST");
                conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
                conn.setDoOutput(true);

                byte[] out = payload.toString().getBytes("UTF-8");
                try (OutputStream os = new BufferedOutputStream(conn.getOutputStream())) {
                    os.write(out);
                    os.flush();
                }

                int status = conn.getResponseCode();
                BufferedReader reader = new BufferedReader(new InputStreamReader(
                        status >= 200 && status < 300 ? conn.getInputStream() : conn.getErrorStream()
                ));
                StringBuilder resp = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    resp.append(line);
                }
                reader.close();

                if (status >= 200 && status < 300) {
                    Log.d("VentasServiceJava", "Webhook enviado correctamente: " + resp.toString());
                    if (callback != null) callback.onResult(true, resp.toString());
                } else {
                    Log.w("VentasServiceJava", "Error webhook HTTP " + status + ": " + resp.toString());
                    if (callback != null) callback.onResult(false, "HTTP " + status + " - " + resp.toString());
                }
            } catch (Exception e) {
                Log.e("VentasServiceJava", "Error enviando webhook", e);
                if (callback != null) callback.onResult(false, e.getMessage() == null ? "Exception" : e.getMessage());
            } finally {
                if (conn != null) conn.disconnect();
            }
        }).start();
    }
}

File: app/src/main/java/com/example/recomendaciones_app/services/VentasService.kt
package com.example.recomendaciones_app.services

import android.util.Log
import android.os.Handler
import android.os.Looper
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Servicio responsable de notificar ventas al webhook de n8n.
 * Provee una API compatible con Java a través de `notificarVentaJava` y `getCurrentFormattedDate`.
 */
class VentasService {

    companion object {
        private const val TAG = "VentasService"
        // Actualizado al webhook Railway provisto por el usuario
        private const val WEBHOOK_URL = "https://primary-production-2873d.up.railway.app/webhook-test/products-sync"

        @JvmStatic
        fun getCurrentFormattedDate(): String {
            val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
            return sdf.format(Date())
        }
    }

    // Data class anidada: desde Java puede instanciarse como `new VentasService.VentaData(...)`.
    data class VentaData(
        val producto_id: Int,
        val producto_nombre: String,
        val cantidad: Int,
        val precio_unitario: Double,
        val total: Double,
        val usuario_id: String?,
        val fecha: String,
        val metodo_pago: String
    )

    // Interfaz para interop con Java
    interface VentaCallbackJava {
        fun onResult(success: Boolean, message: String)
    }

    private val client = OkHttpClient()

    /**
     * Notifica la venta al webhook. Ejecuta el callback en el hilo principal.
     */
    fun notificarVentaJava(venta: VentaData, callback: VentaCallbackJava?) {
        // Construir JSON
        val json = JSONObject().apply {
            put("producto_id", venta.producto_id)
            put("producto_nombre", venta.producto_nombre)
            put("cantidad", venta.cantidad)
            put("precio_unitario", venta.precio_unitario)
            put("total", venta.total)
            put("usuario_id", venta.usuario_id)
            put("fecha", venta.fecha)
            put("metodo_pago", venta.metodo_pago)
        }

        val payload = json.toString()
        Log.d(TAG, "Payload a enviar: $payload")

        val mediaType = "application/json; charset=utf-8".toMediaType()
        val body = payload.toRequestBody(mediaType)
        val request = Request.Builder()
            .url(WEBHOOK_URL)
            .post(body)
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.e(TAG, "Error enviando webhook: ${e.message}", e)
                Handler(Looper.getMainLooper()).post {
                    callback?.onResult(false, "Network error: ${e.message}")
                }
            }

            override fun onResponse(call: Call, response: Response) {
                // Intentar leer el cuerpo de la respuesta para mayor detalle
                val responseBodyStr = try {
                    response.body?.string() ?: ""
                } catch (t: Throwable) {
                    Log.w(TAG, "No se pudo leer response body: ${t.message}")
                    ""
                }

                val success = response.isSuccessful
                val code = try { response.code } catch (_: Throwable) { -1 }
                val msgText = try { response.message } catch (_: Throwable) { "" }

                Log.d(TAG, "HTTP $code - message: $msgText - body: $responseBodyStr")

                val msg = if (success) {
                    "Webhook notified (HTTP $code) - body: $responseBodyStr"
                } else {
                    "Webhook error (HTTP $code): $msgText - body: $responseBodyStr"
                }
                // asegurar callback en hilo principal
                Handler(Looper.getMainLooper()).post {
                    callback?.onResult(success, msg)
                }
                response.close()
            }
        })
    }
}
